technical-docs/K8SSANDRA-OPERATOR-DESIGN.md

```diff
@@ -0,0 +1,410 @@
+
+```
+apiVersion: k8ssandra.io/v1alpha1
+kind: Stargate
+metadata:
+  name: k8ssandra
+spec:
+  replicas: 3
+  cassandraDatacenter:
+    name: dc1
+```
+
+### CassandraBackup / CassandraRestore
+
+These are defined in the [medusa-operator](https://github.com/k8ssandra/medusa-operator) repo. I propose consolidating this into the k8ssandra-operator repo.
+
+See this [GitHub discussion](https://github.com/k8ssandra/k8ssandra/discussions/273) more for ideas around backup/restore and CRDs.
+
+### MonitoringConfiguration
+
+```
+apiVersion: k8ssandra.io/v1alpha1
+kind: MonitoringConfiguration
+metadata:
+  name: k8ssandra
+spec:
+  prometheus:
+    # prometheus properties...
+  grafana:
+  # grafana properties...
+```
+
+With the Grafana license changes and discussion around Victoria Metrics, I think it is a good idea to encapsulate Prometheus and Grafana.
+
+### Grafana, GrafanaDashboard, GrafanaDatasource
+
+Defined in [grafana-operator](https://github.com/integr8ly/grafana-operator).
+
+### Prometheus
+
+Defined in [prometheus-operator](https://github.com/integr8ly/grafana-operator).
+
+## Scenarios
+
+Each scenario consists of a series of events. Each event lists the steps performed by controllers. Controllers run concurrently. When you see multiple controllers listed for an event assume that the steps are performed concurrently and potentially in parallel.
+
+### Create K8ssandraCluster
+
+Let's first consider creating a single DC cluster. User creates a K8ssandraCluster object via kubectl apply:
+
+```
+apiVersion: k8ssandra.io/v1alpha1
+kind: K8ssandraCluster
+metadata:
+  name: test
+spec:
+  cassandra:
+    clusterName: "test"
+    version: "4.0.0" 
+    auth:
+      enabled: true
+    datacenters:
+    - name: dc1
+      size: 3
+  reaper:
+    # reaper properties...
+  stargate:
+    enabled: true
+  medusa:
+    enabled: true
+  monitoring:
+    prometheus:
+      enabled: true
+    grafana:
+      enabled: true
+```
+
+#### Event: K8ssandraCluster created
+
+##### k8ssandra controller
+
+* Start reconciliation loop when K8ssandraCluster object is created
+* Creates a CassandraCluster
+  * Update status
+* Creates a Reaper
+  * Update status
+* Creates a Stargate
+  * Update status
+* Creates a MonitoringConfiguration
+  * Update status
+* End reconciliation loop
+
+#### Event: CassandraCluster created
+
+##### cassandra cluster controller
+
+* Start reconciliation loop when CassandraCluster object is created
+* Create CassandraDatacenter
+  * Update status
+* End reconciliation loop
+
+#### Event: Reaper created
+
+##### reaper controller
+
+* Start reconciliation loop when Reaper object is created
+* Creates reaper Service
+* Requeue request until CassandraDatacenter is ready
+
+#### Event: Stargate created
+
+##### stargate controller
+
+* Start reconciliation loop when Stargate object is created
+* Creates stargate Service
+* Creates ServiceMonitor
+* Creates a GrafanaDashboard
+* Requeue request until CassandraDatacenter is ready
```

Comment:
I feel like these "requeue" until something is ready are unnecessary and cause "ignore-this errors", when we could just create them after we know the resource is ready. It reduces noise in case of deployment error.